src/basedtyping/transformer.py

```diff
@@ -0,0 +1,231 @@
+from __future__ import annotations
+
+import ast
+import types
+import typing
+import uuid
+from contextlib import contextmanager
+from enum import Enum
+from typing import cast
+
+import typing_extensions
+
+import basedtyping
+
+
+# ruff: noqa: N802, S101
+class CringeTransformer(ast.NodeTransformer):
+    """Transforms `1 | 2` into `Literal[1] | Literal[2]` etc"""
+
+    def __init__(
+        self,
+        globalns: dict[str, object] | None,
+        localns: dict[str, object] | None,
+        *,
+        string_literals: bool,
+    ):
+        self.string_literals = string_literals
+
+        # This logic for handling Nones is copied from typing.ForwardRef._evaluate
+        if globalns is None and localns is None:
+            globalns = localns = {}
+        elif globalns is None:
+            # apparently pyright doesn't infer this automatically
+            assert localns is not None
+            globalns = localns
+        elif localns is None:
+            # apparently pyright doesn't infer this automatically
+            assert globalns is not None
```

Comment:
both mypy and pyright narrow it here ([mypy](https://mypy-play.net/?mypy=basedmypy-latest&python=3.12&gist=4b950137a8844a91c65be02170c7a8cc), [pyright](https://basedpyright.com/?typeCheckingMode=all&code=MYGwhgzhAEBiD28BcAoa7oBMCmAzaA%2BgQJYB2xALkQBRob0TYi4A0d96A5iPAEZghSEJFmLAKAbQgUATi2h8AVtnEBdaAB9oAOXilsbDhh7ABQkZjGTpchb2VrNOvQfYcAVIaPQbZTgRBKbBkBYWheRBAvdABKVG8fJlwAOl9Sf0CKYNDoAF4fWT8AoJCQCBQ3egBiaAAVAAtiGB5OMWhceBloerBSTED0530YJuhgeAAHYmxMdpl4AFtoCgBPKfTkhBkAdzAZTAAlPGSCbAA3AQBXMCzKjGJ8bj4zEZhdfWhe2ZMX6FH37DxBIYJ78QQwfI-cF5aAAbwAvnd0EwHtBQb9-i4gcDoDJztgBARVhNsNQoUIYtAapZxFJZPIlCoKOotACkRxIIwZBRoOTXtBSPAeWycVweGChDC%2BeyUfg%2BX83lj2fQ8WcCSAiWtSejwZTqVY6bZGWplRhOcEeTrJaNBcKXKb0PL8laIEA))